misc/peer-store/src/store.rs

```diff
@@ -0,0 +1,99 @@
+use std::{
+    collections::{HashMap, HashSet},
+    time::SystemTime,
+};
+
+use libp2p_core::{Multiaddr, PeerId};
+
+/// A store that
+/// - keep track of currently connected peers;
+/// - contains all observed addresses of peers;
+pub trait Store {
+    /// Called when a peer connects.
+    fn on_peer_connect(&mut self, peer: &PeerId);
+    /// Called when a peer disconnects.
+    fn on_peer_disconnect(&mut self, peer: &PeerId);
+    /// Update an address record.  
+    /// Return `true` when the address is new.  
+    fn on_address_update(&mut self, peer: &PeerId, address: &Multiaddr) -> bool;
+    fn list_connected(&self) -> Box<[&PeerId]>;
+    fn addresses_of_peer(&self, peer: &PeerId) -> Option<Box<[super::AddressRecord]>>;

```

Comment:
Do we need the heap allocation here? Or could these methods instead just return `impl Iterator`?

Reply:
I think it would be better to return `impl Iterator`, unless there some other reason to leave it as is.

Reply:
Well I was really struggling to made the decision. Sure return an iterator why not, since they are both references.